cmd/zoekt-sourcegraph-indexserver/main.go

```
@@ -1049,6 +1076,177 @@ func (s *Server) forceIndex(id uint32) (string, error) {
+		// but the metadata will be empty.
+		if state == indexStateNoop || state == indexStateSuccessMeta || state == indexStateSuccess {
+			repo, metadata, ok, err := args.BuildOptions().FindRepositoryMetadata()
+			if err != nil {
+				errorLog.Printf("failed to read metadata for %s: %s", args.String(), err)
+				return
+			}
+			if !ok {
+				errorLog.Printf("failed to read metadata for %s: failed to find metadata", args.String())
+				return
+			}
+
+			indexTimeUnix = metadata.IndexTime.Unix()
+			zoektRepo = repo
+		}
+
+		switch state {
+		case indexStateSuccess:
+			var branches []string
+			for _, b := range zoektRepo.Branches {
+				branches = append(branches, fmt.Sprintf("%s=%s", b.Name, b.Version))
+			}
+			s.logger.Info("updated index",
+				sglog.Int("tenant", args.TenantID),
+				sglog.String("repo", args.Name),
+				sglog.Uint32("id", args.RepoID),
+				sglog.Strings("branches", branches),
+				sglog.Duration("duration", elapsed),
+			)
+		case indexStateSuccessMeta:
+			infoLog.Printf("updated meta %s in %v", args.String(), elapsed)
+		case indexStateNoop:
+		case indexStateFail:
+			// This should never happen, because indexStateFail implies
+			// indexErr!=nil and we exit early, but we'll handle it gracefully
+			// just in case.
+			indexErr = status.Error(codes.Internal, "failed to index repository")
+		case indexStateEmpty:
+			// Repository exists but is empty, return OK with empty response
+		default:
+			indexErr = status.Errorf(codes.Internal, "unknown index state: %s", state)
+		}
+	})
+
+	if !ran {
+		// Someone else is processing the repository.
+		debugLog.Printf("index job for repository already running: %s", args)
+		return nil, status.Error(codes.AlreadyExists, "repository is already being indexed")
+	}
+
+	if indexErr != nil {
+		return nil, indexErr
+	}
+
+	// Convert branches to proto format
+	repoBranchesProto := make([]*configv1.ZoektRepositoryBranch, len(zoektRepo.Branches))
```

Couldn't `zoektRepo` be nil here? In general, it feels like we shouldn't ignore errors reading metadata, it would just be cleaner (and more accurate) to fail the request, instead of returning an incomplete response.

---

> Couldn't zoektRepo be nil here

Good catch. No objections to failing the request. I wanted to avoid a reindex, but a second index request would be a noop anyway, so that should be fine.